src/bin/stellar-xdr/guess.rs

```diff
@@ -0,0 +1,211 @@
+use std::{
+    cmp,
+    fs::File,
+    io::{self, stdin, Read},
+    path::PathBuf,
+};
+
+use clap::{Args, ValueEnum};
+
+use crate::Channel;
+
+#[derive(thiserror::Error, Debug)]
+#[allow(clippy::enum_variant_names)]
+pub enum Error {
+    #[error("error decoding XDR: {0}")]
+    ReadXdrCurr(#[from] stellar_xdr::curr::Error),
+    #[error("error decoding XDR: {0}")]
+    ReadXdrNext(#[from] stellar_xdr::next::Error),
+    #[error("error reading file: {0}")]
+    ReadFile(#[from] std::io::Error),
+}
+
+#[derive(Args, Debug, Clone)]
+#[command()]
+pub struct Cmd {
+    /// File to decode, or stdin if omitted
+    #[arg()]
+    file: Option<PathBuf>,
+
+    // Input format of the XDR
+    #[arg(long, value_enum, default_value_t)]
+    input: InputFormat,
+
+    // Output format
+    #[arg(long, value_enum, default_value_t)]
+    output: OutputFormat,
+
+    /// Certainty as an arbitrary value
+    #[arg(long, default_value = "2")]
+    certainty: usize,
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, ValueEnum)]
+pub enum InputFormat {
+    Single,
+    SingleBase64,
+    Stream,
+    StreamBase64,
+    StreamFramed,
+}
+
+impl Default for InputFormat {
+    fn default() -> Self {
+        Self::SingleBase64
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, ValueEnum)]
+pub enum OutputFormat {
+    List,
+}
+
+impl Default for OutputFormat {
+    fn default() -> Self {
+        Self::List
+    }
+}
+
+macro_rules! run_x {
+    ($f:ident, $m:ident) => {
+        fn $f(&self) -> Result<(), Error> {
+            let mut f = ResetRead::new(self.file()?);
+            for v in stellar_xdr::$m::TypeVariant::VARIANTS {
+                f.reset();
+                let (ok, err) = match self.input {
+                    InputFormat::Single => stellar_xdr::$m::Type::read_xdr_to_end(v, &mut f)
+                        .map(|_| (1, 0))
+                        .unwrap_or_else(|_| (0, 1)),
+                    InputFormat::SingleBase64 => {
+                        stellar_xdr::$m::Type::read_xdr_base64_to_end(v, &mut f)
+                            .map(|_| (1, 0))
+                            .unwrap_or_else(|_| (0, 1))
+                    }
+                    InputFormat::Stream => stellar_xdr::$m::Type::read_xdr_iter(v, &mut f)
+                        .take(self.certainty)
+                        .fold((0, 0), |c, v| match v {
+                            Ok(_) => (c.0 + 1, c.1),
+                            Err(_) => (c.0, c.1 + 1),
```

Comment:
These lines don't early exit on failure, and it would be better if they did. There's probably a more efficient way to do this, but I didn't want to spend too much time on it so opted for this at the moment.